src/state/opamSysInteract.ml

```diff
@@ -596,6 +598,108 @@ let packages_status ?(env=OpamVariable.Map.empty) packages =
+                 "out": null
+               }
+             }
+           }
+        *)
+        | (`O packages) ->
+            let maybe_pnames = List.map (fun (_key, package_json) ->
+              match package_json with
+              | `O fields -> begin
+                match List.assoc_opt "pname" fields with
+                | Some (`String pname) -> Some pname
+                | _ -> None
+              end
+              | _ -> None) packages
+            in
+            if List.exists Option.is_none maybe_pnames
+            then None
+            else Some (List.filter_map Fun.id maybe_pnames)
+        |  _ -> None
+      in
+      let get_all_available_package_names () =
+        (* Returns a list containing the names of all available packages. We
+           use this to determine which specified packages are available in the
+           repo. The command for querying whether a single package is available
+           is slow, and takes the same amount of time as listing all the
+           packages (~30s). We need to check the availability of multiple
+           packages, so it's faster to list all packages once and then search
+           the output for the specified packages.
+        *)
+        let args = ["--query"; "--available"; "--json"] in
+        let (code, output_lines) = run_command nix_env args in
+        let output = String.concat "\n" output_lines in
+        if code <> 0 then
+          Printf.ksprintf failwith
+            "failed with unexpected exit code %d running command:\n    %s\n\nOutput:\n%s"
+            code (String.concat " " (nix_env::args)) output;
+        match OpamJson.of_string output with
+        | Some json ->
+          begin match all_pnames_of_nix_package_json json with
+            | Some pnames -> pnames
+            | None ->
+              Printf.ksprintf failwith
+                "unexpected format of json string produced by running command:\n    %s\n\nOutput:\n%s"
+                (String.concat " " (nix_env::args)) output
+          end
+        | None ->
+          Printf.ksprintf failwith
+            "failed to parse output as json from running command:\n    %s\n\nOutput:\n%s"
+            (String.concat " " (nix_env::args)) output
+      in
+      let not_installed = OpamSysPkg.Set.filter (Fun.negate is_package_installed) packages in
+      if OpamSysPkg.Set.is_empty not_installed then
+        OpamSysPkg.Set.empty, OpamSysPkg.Set.empty
+      else
+        let not_installed_by_str =
+          OpamSysPkg.Set.elements not_installed
+          |> List.map (fun pkg -> OpamSysPkg.to_string pkg, pkg)
+          |> OpamStd.String.Map.of_list
+        in
+        let available_to_install =
+          get_all_available_package_names ()
+          |> List.filter_map (fun pname -> OpamStd.String.Map.find_opt pname not_installed_by_str)
+          |> OpamSysPkg.Set.of_list
+        in
+        (available_to_install, OpamSysPkg.Set.diff not_installed available_to_install)
```

Comment:
Internal function [compute_sets](https://github.com/ocaml/opam/blob/master/src/state/opamSysInteract.ml#L152) simplifies this computation.